### PR TITLE
Add assign PIN to G-CODE macro feature

### PR DESCRIPTION
**Pull request by MoellerDi**
### Description

This PR is based on the great idea and work from @android444 (PR #17586) and based on the communication between him and @thinkyhead.

This PR might not be ready and should be seen as proof of concept.
(it was tested on my own printer and works fine)

### Benefits

This PR once completed is enabling the user to assign a PIN to predefined g-code macros ~~by using already existing custom menu option~~.

### Related Issues

no issues found but it's related  to PR #17586
maybe related to #7932

### CODE REVIEW

**Comment by android444** [comment]
I don't know why ANYONE who has authority does not accept this method.
If there is a pattern how to do it, I will gladly do it.

**Comment by thinkyhead** [comment]
Deferred until after the next release.

File: Marlin/src/MarlinCore.cpp [hunk 1]
```diff
@@ -1155,79 +1155,79 @@ void setup() {
     #define INIT_CUSTOM_USER_BUTTON_PINS(N) (PIN_EXISTS(USER_GCODE_PIN_##N) && ENABLED(USER_GCODE_PIN_INIT_##N) && defined(USER_GCODE_PIN_STATE_##N))
     #define INIT_CUSTOM_USER_BUTTON_PIN(N) do{ SET_INPUT(USER_GCODE_PIN_##N); WRITE(USER_GCODE_PIN_##N, !USER_GCODE_PIN_STATE_##N); }while(0)
 
-    #if INIT_CUSTOM_USER_BUTTON_PINS(1)
```
Comment (MoellerDi) [inline Marlin/src/MarlinCore.cpp]: The idea here was to allow the user to not init the PIN at boot time (USER_GCODE_PIN_INIT_##N, might be useful if the PIN is already part of another feature) and to ensure proper configuration settings (which are not yet checked in HAS_CUSTOM_USER_BUTTON macro.

**Comment by sjasonsmith** [comment]
What is the state of this? We've certainly had our next release. @MoellerDi do you still consider these changes to be Draft?
I think I will flag this as needing discussion and testing, to see if some other people can try it out.

**Comment by MoellerDi** [comment]
@sjasonsmith: I'll work on it this weekend to get it to the latest code base and do some cleanup. Should be ready for testing by Monday.

@android444: as it was your idea; anything from your end to be added or are you ok to proceed with this PR? Are you ok to get this merged (once ready) or do you prefer to further work on your PR #17586 ?

File: Marlin/src/MarlinCore.cpp (hunk 1 above)
Comment (MoellerDi) [reply Marlin/src/MarlinCore.cpp]: I removed INIT_CUSTOM_USER_BUTTON_PINS(N) macro and USER_GCODE_PIN_INIT_n. That way USER_GCODE_PIN_n will always be initiated to ensure proper state

**Comment by MoellerDi** [comment]
IMHO this PR is now ready for review and testing

**Comment by Zezoh** [comment]
looking forward to test it, :) 

Is it possible to add different scenarios for a single button, to override the limitation of available pins : 

long push 3 sec  
short push 1 sec
one push
two pushes
etc.

**Comment by MoellerDi** [comment]
> looking forward to test it, :)
> 
> Is it possible to add different scenarios for a single button, to override the limitation of available pins :
> 
> long push 3 sec
> short push 1 sec
> one push
> two pushes
> etc.

I'll have a look at that and see what I can do once we are sure the basic functionality works and got tested ;)

**Comment by sjasonsmith** [comment]
I am thinking that this functionality should be separated from Custom Menus. This would allow them to be used from screen-less printers, or printers with screens incompatible with custom menus.

A huge number of printers now use TFT displays or operate from OctoPrint alone. We should make this available for those printers.

**Comment by MoellerDi** [comment]
I agree, this functionality should be separated from Custom Menus in order to support screen-less printers or printers with screens incompatible with custom menus. I will update this PR after xmas holidays.

**Comment by Cynopolis** [comment]
I tried out this pull today on my own build of marlin and it works very well. My specific application for this feature actually uses an analog input. This code had no problem dealing with analog input and registered a logic level HIGH at 2-3V. I was also surprised to see that it can use data in pins that are also being used by other parts of the Marlin firmware.

**Comment by MoellerDi** [comment]
this PR is now separated from Custom Menus and works fine on my printer. IMHO it should be ready for prime time but let me know if something is needed from my end

**Comment by DrumClock** [comment]
Hi, @MoellerDi 
I wanted to define `CUSTOM_USER_BUTTONS` but somehow it doesn't work. 

CUB

PC10


Where is the mistake?

**Comment by DrumClock** [comment]
if I disable the `#if` condition, it's functional.  `(BNT_PIN = PC10)`

UB2

**Comment by ellensp** [comment]
himm, that #if looks wrong.
from macros
#define PIN_EXISTS(PN)      (defined(PN##_PIN) && PN##_PIN >= 0)

which means it appends the _PIN to the name,  so it should be #if PIN_EXISTS(BUTTON1)

**Comment by DrumClock** [comment]
Hi @ellensp 
yes `#if PIN_EXISTS (BUTTON1)` already works

Should I fix all BUTTON conditions?

**Comment by ellensp** [comment]
yes, are you going to create PR? I will If you don't.

**Comment by DrumClock** [comment]
...unfortunately I can't create PR.

**Comment by ellensp** [comment]
ok ill do it.

**Comment by ellensp** [comment]
PR to fix this has been created

**Pull request merged**